bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,75 +1,216 @@
+            return None;
+        }
+
+        let failed_attempts = self.retries_amount();
+
+        let worker_status = match self.status_client.status(StatusRequest {}).await {
+            Ok(response) => response.into_inner(),
+            Err(e) => {
+                error!("Failed to check worker status ({}): {}", self.address(), e);
+                return Some(WorkerHealthStatus::Unhealthy {
+                    failed_attempts: failed_attempts + 1,
+                    first_fail_timestamp: Instant::now(),
+                    reason: e.message().to_string(),
+                });
             },
+        };
+
+        self.version = worker_status.version;
+
+        let worker_supported_proof_type =
+            match ProverType::try_from(worker_status.supported_proof_type) {
+                Ok(proof_type) => proof_type,
+                Err(e) => {
+                    error!(
+                        "Failed to convert worker supported proof type ({}): {}",
+                        self.address(),
+                        e
+                    );
+                    return Some(WorkerHealthStatus::Unhealthy {
+                        failed_attempts: failed_attempts + 1,
+                        first_fail_timestamp: Instant::now(),
+                        reason: e.to_string(),
+                    });
+                },
+            };
+
+        if !(*supported_prover_type == worker_supported_proof_type) {
+            return Some(WorkerHealthStatus::Unhealthy {
+                failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Unsupported proof type".to_string(),
+            });
+        }
+
+        if !worker_status.ready {
+            return Some(WorkerHealthStatus::Unhealthy {
+                failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Worker not ready".to_string(),
+            });
         }
+
+        Some(WorkerHealthStatus::Healthy)
     }
 
+    /// Returns the worker availability.
     pub fn is_available(&self) -> bool {
         self.is_available
     }
 
+    /// Sets the worker availability.
     pub fn set_availability(&mut self, is_available: bool) {
-        self.is_available = is_available;
+        self.is_available = is_available
+    }
+
+    pub(crate) fn set_health_status(&mut self, health_status: WorkerHealthStatus) {
+        self.health_status = health_status;
+        match &self.health_status {
+            WorkerHealthStatus::Healthy => {
+                self.is_available = true;
+            },
+            WorkerHealthStatus::Unhealthy { .. } => {
+                WORKER_UNHEALTHY.with_label_values(&[&self.address()]).inc();
+                self.is_available = false;
+            },
+        }
+    }
+
+    /// Returns the number of retries the worker has had.
+    pub fn retries_amount(&self) -> usize {
```

Comment:
I would rename this to `num_retries`.

Reply:
I think something like `failed_attempts` or `num_failures`/`num_failed_checks` to follow the pattern suggested above would be better. "Retries" is a bit misleading here, since a single failed attempt would imply it was not retried yet, but we still would return `1`.

Reply:
Yes, agreed. I like `num_failures` or `num_failed_checks`.